identity_stardust/src/client/client_ext.rs

```diff
@@ -0,0 +1,616 @@
+use iota_client::block::output::feature::SenderFeature;
+use iota_client::block::output::unlock_condition::AddressUnlockCondition;
+use iota_client::block::output::unlock_condition::GovernorAddressUnlockCondition;
+use iota_client::block::output::unlock_condition::StateControllerAddressUnlockCondition;
+use iota_client::block::output::AliasId;
+use iota_client::block::output::AliasOutput;
+use iota_client::block::output::AliasOutputBuilder;
+use iota_client::block::output::BasicOutputBuilder;
+use iota_client::block::output::Feature;
+use iota_client::block::output::Output;
+use iota_client::block::output::OutputId;
+use iota_client::block::output::RentStructure;
+use iota_client::block::output::UnlockCondition;
+use iota_client::block::payload::transaction::TransactionEssence;
+use iota_client::block::payload::Payload;
+use iota_client::block::Block;
+use iota_client::secret::SecretManager;
+use iota_client::Client;
+
+use crate::error::Result;
+use crate::Error;
+use crate::NetworkName;
+use crate::StardustDID;
+use crate::StardustDocument;
+
+/// An extension trait for a [`Client`] that provides helper functions for publication
+/// and resolution of DID documents in Alias Outputs.
+///
+/// This trait is only meant to be used rather than implemented.
+#[async_trait::async_trait]
+pub trait StardustClientExt: Sync {
+  /// Returns a reference to a [`Client`].
+  fn client(&self) -> &Client;
+
+  /// Create a DID with a new Alias Output containing the given `document`.
+  ///
+  /// The `address` will be set as the state controller and governor unlock conditions.
+  /// The minimum required token deposit amount will be set according to the given
+  /// `rent_structure`, which will be fetched from the node if not provided.
+  /// The returned Alias Output can be further customized before publication, if desired.
+  ///
+  /// NOTE: this does *not* publish the Alias Output. See [`publish_did_output`](StardustClientExt::publish_did_output).
+  ///
+  /// # Errors
+  ///
+  /// - Returns an [`Error::DIDUpdateError`] when retrieving the `RentStructure` fails.
+  /// - Returns an [`Error::AliasOutputBuildError`] when building the Alias Output fails.
+  async fn new_did(
+    &self,
+    address: Address,
+    document: StardustDocument,
+    rent_structure: Option<RentStructure>,
+  ) -> Result<AliasOutput> {
+    let rent_structure: RentStructure = if let Some(inner) = rent_structure {
+      inner
+    } else {
+      self
+        .client()
+        .get_rent_structure()
+        .await
+        .map_err(Error::DIDUpdateError)?
+    };
+
+    AliasOutputBuilder::new_with_minimum_storage_deposit(rent_structure, AliasId::null())
+      .map_err(Error::AliasOutputBuildError)?
+      .with_state_index(0)
+      .with_foundry_counter(0)
+      .with_state_metadata(document.pack()?)
+      .add_feature(Feature::Sender(SenderFeature::new(address)))
+      .add_immutable_feature(Feature::Issuer(IssuerFeature::new(address)))
```

Comment:
The issuer feature could be undesirable or seen as unnecessary. I suppose people could remove it themselves though, if possible.

Reply:
Removed it.

According to TIP-18 it's primary use case is:

> [...] proving authenticity of NFTs. Whenever an NFT is minted as an NFT output, the creator (issuer) can fill the Issuer Feature with their address that they have to unlock in the transaction. Issuers then can publicly disclose their addresses to prove the authenticity of the NFT once it is in circulation.